### PR TITLE
feature(integration): add track call support for freshmarketer and freshsales

### PR DESCRIPTION
**Pull request by sauravlal-rudderstack**
## Description of the change

> add track call support for freshmarketer and freshsales

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/freshmarketer/utils.js [hunk 1]
```diff
@@ -46,36 +90,316 @@ const getUserAccountDetails = async (userEmail, Config) => {
...
+      }
+    ];
+  } else {
+    accountDetails = [
+      {
+        id: accountId,
+        is_primary: true
+      }
+    ];
+  }
```
Comment (sandeepdsvs) [inline v0/destinations/freshmarketer/utils.js:117]:
```suggestion
  const accountDetail = {
    id: accountId,
    is_primary: false
  };
  if (accountDetails.length > 0) {
    accountDetails = [...accountDetails, accountDetail];
  } else {
    accountDetail.is_primary = true;
    accountDetails = [accountDetail];
  }
```

File: v0/destinations/freshmarketer/utils.js [hunk 2]
```diff
@@ -46,36 +90,316 @@ const getUserAccountDetails = async (userEmail, Config) => {
       errorStatus
     );
   }
-  return userSalesAccountResponse;
+  let accountDetails =
+    userSalesAccountResponse.response.contact?.sales_accounts;
```
Comment (sandeepdsvs) [inline v0/destinations/freshmarketer/utils.js:94]:
```suggestion
    userSalesAccountResponse?.response?.contact?.sales_accounts;
```

Comment (sandeepdsvs) [reply v0/destinations/freshmarketer/utils.js:94]: Do the same at other possible places as well

Comment (sauravlal-rudderstack) [reply v0/destinations/freshmarketer/utils.js:94]: okay

File: v0/destinations/freshmarketer/utils.js (hunk 1 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshmarketer/utils.js:117]: Okay, will update

File: v0/destinations/freshsales/utils.js [hunk 3]
```diff
@@ -39,44 +81,284 @@ const getUserAccountDetails = async (userEmail, Config) => {
...
+    }
+  };
+  if (!payload.sales_activity_name && !payload.sales_activity_type_id) {
+    throw new CustomError(
+      `Either of sales activity name or sales activity type id is required. Aborting!`,
+      400
+    );
+  }
+
+  const email = get(message, "email");
```
Comment (utsabc) [inline v0/destinations/freshsales/utils.js:195]: Suggestion
```
const email = getFieldValueFromMessage(message, "email");
```

File: v0/destinations/freshsales/transform.js [hunk 4]
```diff
@@ -48,6 +66,54 @@ const identifyResponseBuilder = (message, { Config }) => {
...
+ * @param {*} Config
+ * @returns
+ */
+const trackResponseBuilder = async (message, { Config }) => {
+  const { event } = message;
+  if (!event) {
+    throw new CustomError("Event name is required for track call.", 400);
+  }
+  let payload;
+  const email = getFieldValueFromMessage(message, "email");
```
Comment (utsabc) [inline v0/destinations/freshsales/transform.js:81]: This is not used

File: v0/destinations/freshsales/utils.js [hunk 5]
```diff
@@ -39,44 +81,284 @@ const getUserAccountDetails = async (userEmail, Config) => {
...
+
+  if (!payload.targetable_id && !email) {
```
Comment (utsabc) [inline v0/destinations/freshsales/utils.js:197]: targetable_id* 	number 	ID of the contact/deal/account related to the sales activity created

looks like `targetable_id` is an internal id of freshsales shouldn't we map this to our externalId?

In rudderPayload example: 
.....
externalId: {
    type: "freshworks_contact"
    id: 123
}

File: v0/destinations/freshsales/utils.js [hunk 6]
```diff
@@ -39,44 +81,284 @@ const getUserAccountDetails = async (userEmail, Config) => {
...
+  email,
+  Config,
+  payload,
+  salesActivityTypeId
+) => {
+  const userDetails = await getContactsDetails(email, Config);
+  const userId = userDetails.response.contact.id;
+  const responseBody = {
+    ...payload,
+    targetable_id: userId,
```
Comment (utsabc) [inline v0/destinations/freshsales/utils.js:168]: Directly set 
```
targetable_id: userDetails.response.contact.id
```

File: v0/destinations/freshsales/utils.js [hunk 7]
```diff
@@ -39,44 +81,284 @@ const getUserAccountDetails = async (userEmail, Config) => {
...
+    throw new CustomError(
+      `failed to fetch sales activities details ${errMessage}`,
+      errorStatus
+    );
+  }
+
+  const salesActivityList = salesActivityResponse.response.sales_activity_types;
+  if (salesActivityList && Array.isArray(salesActivityList)) {
+    const listDetails = salesActivityList.find(
+      list => list.name === payload.sales_activity_name
```
Comment (utsabc) [inline v0/destinations/freshsales/utils.js:237]:
```
salesActivity => salesActivity.name === payload.sales_activity_name
```

File: v0/destinations/freshsales/utils.js [hunk 8]
```diff
@@ -39,44 +81,284 @@ const getUserAccountDetails = async (userEmail, Config) => {
...
+    const errorStatus = salesActivityResponse.response.errors?.code || "500";
+    throw new CustomError(
+      `failed to fetch sales activities details ${errMessage}`,
+      errorStatus
+    );
+  }
+
+  const salesActivityList = salesActivityResponse.response.sales_activity_types;
+  if (salesActivityList && Array.isArray(salesActivityList)) {
+    const listDetails = salesActivityList.find(
```
Comment (utsabc) [inline v0/destinations/freshsales/utils.js:236]:
```
const salesActivityDetails
```

File: v0/destinations/freshsales/utils.js [hunk 9]
```diff
@@ -39,44 +81,284 @@ const getUserAccountDetails = async (userEmail, Config) => {
...
+        email,
+        Config,
+        payload,
+        listDetails.id
+      );
+    }
+
+    return responseBody;
+  }
+  throw new CustomError(`sales activity types must be an array.`, 400);
```
Comment (utsabc) [inline v0/destinations/freshsales/utils.js:263]: Unable to fetch correct list of sales activity

File: v0/destinations/freshsales/utils.js [hunk 10]
```diff
@@ -39,44 +81,284 @@ const getUserAccountDetails = async (userEmail, Config) => {
...
+    );
+  }
+  const lifecycleStageId = getFieldValueFromMessage(
+    message,
+    "lifecycleStageId"
+  );
+  const lifecycleStageName = getFieldValueFromMessage(
+    message,
+    "lifecycleStageName"
+  );
```
Comment (utsabc) [inline v0/destinations/freshsales/utils.js:292]:
```
lifecycleStageId, and lifecycleStageName are not present in GenericFieldMapping.json
```
You have to use example `get(message, "properties.lifecycleStageName")`

File: v0/destinations/freshmarketer/transform.js [hunk 11]
```diff
@@ -7,16 +8,34 @@ const {
...
 } = require("../../util");
 
-const { CONFIG_CATEGORIES, MAPPING_CONFIG } = require("./config");
 const {
-  createUpdateAccount,
+  CONFIG_CATEGORIES,
+  MAPPING_CONFIG,
+  DESTINATION,
+  IDENTIFY_MAX_BATCH_SIZE,
+  BATCH_IDENTIFY_ENDPOINT
```
Comment (sandeepdsvs) [inline v0/destinations/freshmarketer/transform.js:21]: Not being used anywhere

File: v0/destinations/freshsales/utils.js [hunk 12]
```diff
@@ -39,44 +81,284 @@ const getUserAccountDetails = async (userEmail, Config) => {
...
+    throw new CustomError(
+      `failed to fetch lifecycle_stages details ${errMessage}`,
+      errorStatus
+    );
+  }
+  const lifeCycleStages = lifeCycleStagesResponse.response?.lifecycle_stages;
+  if (lifeCycleStages && Array.isArray(lifeCycleStages)) {
+    const listDetails = lifeCycleStages.find(
+      list => list.name === lifecycleStageName
+    );
```
Comment (utsabc) [inline v0/destinations/freshsales/utils.js:323]:
```
 const lifeCycleStageDetials = lifeCycleStages.find(
      lifeCycleStage => lifeCycleStages.name === lifecycleStageName
    );
```

File: v0/destinations/freshsales/transform.js [hunk 13]
```diff
@@ -48,6 +66,54 @@ const identifyResponseBuilder = (message, { Config }) => {
...
+  const response = defaultRequestConfig();
+  switch (
+    event
+      .toLowerCase()
+      .trim()
+      .replace(/\s+/g, "_")
+  ) {
+    case "sales_activity":
```
Comment (utsabc) [inline v0/destinations/freshsales/transform.js:89]: Can we add mapping in UI 

Rudder Event  --->  Freshsales sales_activity or lifecycle_stage (drop down)

File: v0/destinations/freshmarketer/deleteUsers.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        .build();
+    }
+  }
+  return { statusCode: 200, status: "successful" };
+};
+
+const processDeleteUsers = event => {
+  const { userAttributes, config } = event;
+  const resp = userDeletionHandler(userAttributes, config);
+  return resp;
```
Comment (sandeepdsvs) [inline v0/destinations/freshmarketer/deleteUsers.js:62]:
```suggestion
  return userDeletionHandler(userAttributes, config);
```

File: v0/destinations/freshsales/utils.js (hunk 6 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshsales/utils.js:171]: updated with some validation.

File: v0/destinations/freshsales/utils.js (hunk 7 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshsales/utils.js]: updated

File: v0/destinations/freshsales/utils.js (hunk 8 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshsales/utils.js]: updated

File: v0/destinations/freshsales/transform.js [hunk 15]
```diff
@@ -65,51 +131,19 @@ const groupResponseBuilder = async (message, { Config }) => {
...
   if (!userEmail) {
-    const response = defaultRequestConfig();
-    response.endpoint = `https://${Config.domain}${CONFIG_CATEGORIES.GROUP.baseUrl}`;
-    response.method = defaultPostRequestConfig.requestMethod;
-    response.body.JSON = payloadBody;
-    response.headers = {
-      Authorization: `Token token=${Config.apiKey}`,
-      "Content-Type": "application/json"
-    };
+    const response = updateAccountWOContact(payload, Config);
```
[the remark this answers is not in the document]
Comment (sauravlal-rudderstack) [reply v0/destinations/freshsales/transform.js:141]: updated

File: v0/destinations/freshsales/utils.js (hunk 9 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshsales/utils.js]: updated

File: v0/destinations/freshsales/utils.js (hunk 10 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshsales/utils.js]: updated

File: v0/destinations/freshmarketer/transform.js [hunk 16]
```diff
@@ -7,16 +8,34 @@ const {
   defaultPostRequestConfig,
   getErrorRespEvents,
   getSuccessRespEvents,
-  getFieldValueFromMessage
+  getFieldValueFromMessage,
+  defaultBatchRequestConfig,
+  handleRtTfSingleEventError
```
[the remark this answers is not in the document]
Comment (sauravlal-rudderstack) [reply v0/destinations/freshmarketer/transform.js]: updated

File: v0/destinations/freshmarketer/transform.js (hunk 11 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshmarketer/transform.js]: updated

File: v0/destinations/freshsales/utils.js (hunk 12 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshsales/utils.js]: updated

File: v0/destinations/freshsales/utils.js (hunk 3 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshsales/utils.js]: updated

File: v0/destinations/freshsales/transform.js (hunk 13 above)
Comment (utsabc) [reply v0/destinations/freshsales/transform.js]: I meant a UI mapping similar to our ads destination check SnapChat conversions for example

Comment (sauravlal-rudderstack) [reply v0/destinations/freshsales/transform.js]: Done

File: v0/destinations/freshmarketer/deleteUsers.js (hunk 14 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshmarketer/deleteUsers.js]: Removed

File: v0/destinations/freshsales/utils.js (hunk 5 above)
Comment (sauravlal-rudderstack) [reply v0/destinations/freshsales/utils.js:206]: Done

**Review by sandeepdsvs — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**